include/ensmallen_bits/eve/eve_impl.hpp

```diff
@@ -0,0 +1,171 @@
+/**
+ * @file eve_impl.hpp
+ * @author Marcus Edel
+ *
+ * Implementation of Eve: a gradient based optimization method with Locally
+ * and globally adaptive learning rates.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_EVE_EVE_IMPL_HPP
+#define ENSMALLEN_EVE_EVE_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "eve.hpp"
+
+#include <ensmallen_bits/function.hpp>
+
+namespace ens {
+
+inline EVE::EVE(const double stepSize,
+                const size_t batchSize,
+                const double beta1,
+                const double beta2,
+                const double beta3,
+                const double epsilon,
+                const double clip,
+                const size_t maxIterations,
+                const double tolerance,
+                const bool shuffle) :
+    stepSize(stepSize),
+    batchSize(batchSize),
+    beta1(beta1),
+    beta2(beta2),
+    beta3(beta3),
+    epsilon(epsilon),
+    clip(clip),
+    maxIterations(maxIterations),
+    tolerance(tolerance),
+    shuffle(shuffle)
+{ /* Nothing to do. */ }
+
+//! Optimize the function (minimize).
+template<typename DecomposableFunctionType>
+double EVE::Optimize(
```

Comment:
Ahh, too bad we could not reuse SGD here.  I guess this is because the Eve update rule requires the function value itself.